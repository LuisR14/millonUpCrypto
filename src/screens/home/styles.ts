import {StyleSheet} from 'react-native';
import {dimension} from '../../themes/dimension';
import colors from '../../themes/colors';
import fontSize from '../../themes/fontSize';
const {horizontalScale, verticalScale} = dimension();

export default StyleSheet.create({
  container: {
    flex: 1,
  },
  textInputContainer: {
    height: verticalScale(40),
    marginTop: 20,
    borderRadius: 8,
    flexDirection: 'row',
    width: horizontalScale(330),
    backgroundColor: colors.lightGray,
    paddingHorizontal: 10,
    paddingVertical: 5,
    alignSelf: 'center',
  },
  textInput: {
    flex: 1,
    paddingHorizontal: 10,
    paddingVertical: 5,
    color: colors.black,
    fontSize: fontSize.neutral,
  },
  filtersContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    marginHorizontal: 25,
  },
  textFilters: {
    padding: 10,
    textAlign: 'center',
    fontSize: fontSize.normal,
    fontWeight: 'bold',
    color: colors.black,
  },
  filtersButton: {
    width: horizontalScale(100),
    borderRadius: 10,
    backgroundColor: colors.lightGray,
  },
  touchable: {
    width: horizontalScale(330),
    borderRadius: 10,
  },
  touchableIconSearch: {
    alignSelf: 'center',
  },
  flatlistContainer: {
    alignItems: 'center',
  },
});
