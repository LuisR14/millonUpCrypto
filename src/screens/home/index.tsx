/* eslint-disable react-hooks/exhaustive-deps */
import React, {useCallback, useEffect, useRef, useState} from 'react';
import {
  Modal,
  RefreshControl,
  Text,
  TextInput,
  TouchableHighlight,
  TouchableOpacity,
  View,
} from 'react-native';
import {CryptoProps} from '../../types/crypto';
import styles from './styles';
import {RenderFooter} from '../../components/flatlistFooter';
import CryptoComponent from '../../components/cryptoList';
import Margin from '../../components/margin';
import SearchIcon from '../../assets/icons/search.svg';
import CleanSearchIcon from '../../assets/icons/cleanSearch.svg';
import colors from '../../themes/colors';
import {getCryptoList} from '../../functions/getCryptoList';
import {loadMoreCrypto} from '../../functions/loadMoreCrypto';
import {searchText} from '../../functions/searchText';
import {SkeletonLoader} from '../../components/skeletonLoader';
import {ModalDetailComponent} from '../../components/modalDetail';
import {useDispatch, useSelector} from 'react-redux';
import {
  create,
  createBackup,
  update,
  updateBackup,
} from '../../redux/cryptoListSlice';
import {CryptoListEmpty} from '../../components/cryptoListEmpty';
import {sleep} from '../../functions/sleep';
import Animated, {
  useAnimatedScrollHandler,
  useSharedValue,
} from 'react-native-reanimated';
import {SafeAreaView} from 'react-native';
import SplashScreen from 'react-native-splash-screen';
import {filter} from '../../functions/filterCryptoList';
import {saveCryptoList} from '../..//functions/saveCryptoList';
import {getCryptoListSaved} from '../..//functions/getCryptoListSaved';
import {useNetInfo} from '@react-native-community/netinfo';
import {useToast} from 'react-native-toast-notifications';

export const HomeScreen = () => {
  const [cryptoDetail, setcryptoDetail] = useState<CryptoProps>();
  const [isLoading, setisLoading] = useState(false);
  const [isRefreshing, setisRefreshing] = useState(false);
  const [isLoadingMore, setisLoadingMore] = useState(false);
  const [count, setcount] = useState(1);
  const [search, setsearch] = useState('');
  const [modalDetail, setmodalDetail] = useState(false);
  const [blockUpdate, setblockUpdate] = useState(true);
  const [activefilterGainer, setActivefilterGainer] = useState(false);
  const [activefilterLosers, setActivefilterLosers] = useState(false);
  const [activefilterVolume24, setActivefilterVolume24] = useState(false);
  const [errorMsg, setErrorMsg] = useState(null);
  const dispatch = useDispatch();
  const searchRef = useRef<any>();
  const flatListRef = useRef<any>();
  const cryptoList = useSelector((state: any) => state.cryptoList);
  const Settings = useSelector((state: any) => state.userSettings);
  const Theme = Settings?.darkMode;
  const netInfo = useNetInfo();
  const Offline = Settings?.offline;
  const scrollY = useSharedValue(0);
  const toast = useToast();
  useEffect(() => {
    SplashScreen?.hide();
    setisLoading(true);
  }, []);

  useEffect(() => {
    if (netInfo.isConnected !== null) {
      checkCryptoApi();
    }
  }, [netInfo.isConnected]);

  useEffect(() => {
    if (Offline) {
      setblockUpdate(true);
      getSavedCryptoList();
    } else if (!Offline && netInfo.isConnected !== null) {
      setblockUpdate(false);
      checkCryptoApi();
    }
  }, [Offline]);

  useEffect(() => {
    if (errorMsg) {
      toast.show(errorMsg, {
        type: 'danger',
      });
      setErrorMsg(null);
    }
  }, [toast, errorMsg]);

  const checkCryptoApi = async () => {
    cleanFilter();
    if (Offline || netInfo.isConnected === false) {
      getSavedCryptoList();
    } else {
      const cryptoListResponse = await getCryptoList({
        setsearch,
      });
      if (cryptoListResponse?.length > 0) {
        dispatch(create(cryptoListResponse));
        dispatch(createBackup(cryptoListResponse));
        saveCryptoList(cryptoListResponse);
        setblockUpdate(false);
      } else {
        setErrorMsg('server error');
      }
    }
    setisLoading(false);
  };
  const updateCryptoList = async () => {
    const cryptoListUpdate = await loadMoreCrypto({
      setisLoadingMore,
      count,
      setcount,
      search,
    });
    if (cryptoListUpdate.length > 0) {
      dispatch(update(cryptoListUpdate));
      dispatch(updateBackup(cryptoListUpdate));
    } else {
      setErrorMsg('server error');
    }
  };

  const getSavedCryptoList = async () => {
    const list = await getCryptoListSaved();
    dispatch(create(list));
    dispatch(createBackup(list));
  };

  const handler = useAnimatedScrollHandler({
    onScroll: event => {
      scrollY.value = event.contentOffset.y;
    },
  });

  const scollToTop = () => {
    flatListRef.current.scrollToOffset({
      animated: true,
      offset: 0,
    });
  };

  const cleanFilter = () => {
    setActivefilterGainer(false);
    setActivefilterLosers(false);
    setActivefilterVolume24(false);
  };

  const keyExtractor = useCallback((item: CryptoProps) => item.id, []);

  const renderItems = useCallback(
    ({item, index}: any) => (
      <TouchableHighlight
        testID="buttonDetailsTest"
        style={styles.touchable}
        activeOpacity={0.6}
        underlayColor="lightgray"
        onPress={() => {
          setcryptoDetail(item);
          setmodalDetail(true);
        }}>
        <CryptoComponent cryptoItem={item} scrollY={scrollY} index={index} />
      </TouchableHighlight>
    ),
    [],
  );

  return (
    <>
      <SafeAreaView style={styles.container}>
        {isLoading && !isRefreshing ? (
          <SkeletonLoader />
        ) : (
          <>
            <View testID="viewTextInputTest" style={styles.textInputContainer}>
              <TouchableOpacity
                style={styles.touchableIconSearch}
                onPress={() => searchRef.current.focus()}>
                <SearchIcon />
              </TouchableOpacity>
              <TextInput
                testID="textInputTest"
                ref={searchRef}
                style={styles.textInput}
                placeholder="Search coin"
                placeholderTextColor={colors.gray}
                value={search}
                onChangeText={text => {
                  scollToTop();
                  cleanFilter();
                  const cryptoListFilter = searchText({
                    text,
                    setsearch,
                    cryptoList,
                  });
                  if (cryptoListFilter) {
                    dispatch(create(cryptoListFilter));
                  } else {
                    dispatch(create(cryptoList?.listBackup));
                  }
                }}
              />
              {search.length > 0 && (
                <TouchableOpacity
                  style={styles.touchableIconSearch}
                  onPress={() => {
                    setsearch('');
                    dispatch(create(cryptoList?.listBackup));
                  }}>
                  <CleanSearchIcon />
                </TouchableOpacity>
              )}
            </View>
            <Margin bottom={15} />
            <View style={styles.filtersContainer}>
              <TouchableOpacity
                testID="buttonFilterGainers"
                onPress={() => {
                  scollToTop();
                  if (activefilterGainer) {
                    setblockUpdate(false);
                    dispatch(create(cryptoList?.listBackup));
                  } else {
                    setActivefilterLosers(false);
                    setActivefilterVolume24(false);
                    setblockUpdate(true);
                    const filterlist = filter(
                      cryptoList?.listBackup,
                      'gainers',
                    );
                    dispatch(create(filterlist));
                  }
                  setActivefilterGainer(!activefilterGainer);
                }}
                style={{
                  ...styles.filtersButton,
                  backgroundColor: activefilterGainer
                    ? Theme === 'dark'
                      ? colors.white
                      : colors.lightBlue
                    : Theme === 'dark'
                    ? colors.gray
                    : colors.lightGray,
                }}>
                <Text style={styles.textFilters}>Gainers</Text>
              </TouchableOpacity>
              <TouchableOpacity
                testID="buttonFilterLosers"
                onPress={() => {
                  scollToTop();
                  if (activefilterLosers) {
                    setblockUpdate(false);
                    dispatch(create(cryptoList?.listBackup));
                  } else {
                    setActivefilterGainer(false);
                    setActivefilterVolume24(false);
                    setblockUpdate(true);
                    const filterlist = filter(cryptoList?.listBackup, 'losers');
                    dispatch(create(filterlist));
                  }
                  setActivefilterLosers(!activefilterLosers);
                }}
                style={{
                  ...styles.filtersButton,
                  backgroundColor: activefilterLosers
                    ? Theme === 'dark'
                      ? colors.white
                      : colors.lightBlue
                    : Theme === 'dark'
                    ? colors.gray
                    : colors.lightGray,
                }}>
                <Text style={styles.textFilters}>Losers</Text>
              </TouchableOpacity>
              <TouchableOpacity
                testID="buttonFilterVolume24"
                onPress={() => {
                  scollToTop();
                  if (activefilterVolume24) {
                    setblockUpdate(false);
                    dispatch(create(cryptoList?.listBackup));
                  } else {
                    setActivefilterLosers(false);
                    setActivefilterGainer(false);
                    setblockUpdate(true);
                    const filterlist = filter(
                      cryptoList?.listBackup,
                      'gainers',
                    );
                    dispatch(create(filterlist));
                  }
                  setActivefilterVolume24(!activefilterVolume24);
                }}
                style={{
                  ...styles.filtersButton,
                  backgroundColor: activefilterVolume24
                    ? Theme === 'dark'
                      ? colors.white
                      : colors.lightBlue
                    : Theme === 'dark'
                    ? colors.gray
                    : colors.lightGray,
                }}>
                <Text style={styles.textFilters}>Volume24h</Text>
              </TouchableOpacity>
            </View>
            <Margin bottom={10} />
            <Animated.FlatList
              testID="flatListTest"
              ref={flatListRef}
              contentContainerStyle={styles.flatlistContainer}
              keyboardDismissMode="on-drag"
              ListEmptyComponent={CryptoListEmpty}
              onScroll={handler}
              maxToRenderPerBatch={40}
              onEndReached={async () => {
                if (!blockUpdate) {
                  setblockUpdate(true);
                  updateCryptoList();
                  await sleep(5000);
                  setblockUpdate(false);
                }
              }}
              onEndReachedThreshold={1}
              refreshControl={
                <RefreshControl
                  refreshing={isRefreshing}
                  onRefresh={() => {
                    setisRefreshing(true);
                    setTimeout(() => {
                      setisRefreshing(false);
                    }, 400);
                    checkCryptoApi();
                  }}
                  title="Refreshing"
                  tintColor={Theme === 'dark' ? colors.white : colors.black}
                  titleColor={Theme === 'dark' ? colors.white : colors.black}
                  colors={
                    Theme === 'dark' ? [colors.lightBlue] : [colors.black]
                  }
                />
              }
              data={cryptoList?.list}
              renderItem={renderItems}
              keyExtractor={keyExtractor}
              ListFooterComponent={() => RenderFooter(isLoadingMore)}
            />
            <Modal
              testID="modalTest"
              style={{
                backgroundColor:
                  Theme !== 'dark' ? colors.white : colors.darkCyan,
              }}
              animationType="slide"
              visible={modalDetail}>
              <ModalDetailComponent
                cryptoItem={cryptoDetail}
                setmodalDetail={setmodalDetail}
              />
            </Modal>
          </>
        )}
      </SafeAreaView>
    </>
  );
};
