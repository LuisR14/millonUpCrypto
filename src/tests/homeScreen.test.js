import React from 'react';
import {HomeScreen} from '../screens/home/index';
import {renderWithProviders} from './utils_for_tests';
import {NavigationContainer} from '@react-navigation/native';
import {cleanup, screen} from '@testing-library/react-native';
import axios from 'axios';

jest.mock('react-native-toast-notifications', () => {
  return {
    useToast: () => ({
      show: jest.fn(),
    }),
  };
});
const mockedNavigate = jest.fn();
const mockedgoBack = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      navigate: mockedNavigate,
      goBack: mockedgoBack,
    }),
  };
});

describe('HomeScreen', () => {
  let component;
  afterEach(cleanup);
  beforeEach(() => {
    component = renderWithProviders(
      <NavigationContainer>
        <HomeScreen />
      </NavigationContainer>,
    );
  });

  it('snapshot of screen to JSON', () => {
    const {toJSON} = renderWithProviders(
      <NavigationContainer>
        <HomeScreen />
      </NavigationContainer>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly home screen', () => {
    expect(component).toBeTruthy();
  });

  it('renders correctly the textInput view', async () => {
    axios.get.mockResolvedValueOnce({
      data: {
        results: [
          {
            id: '90',
            symbol: 'BTC',
            name: 'Bitcoin',
            nameid: 'bitcoin',
            rank: 1,
            price_usd: '25898.77',
            percent_change_24h: '-1.43',
            percent_change_1h: '-0.03',
            percent_change_7d: '-0.78',
            price_btc: '1.00',
            market_cap_usd: '504273572482.94',
            volume24: 7452436883.073213,
            volume24a: 8300378920.653947,
            csupply: '19470946.00',
            tsupply: '19470946',
            msupply: '21000000',
          },
        ],
      },
    });
    const item = screen.getByTestId('viewTextInputTest');
    expect(item).toBeOnTheScreen();
  });

  it('renders correctly the textInput', () => {
    const item = screen.getByTestId('textInputTest');
    expect(item).toBeOnTheScreen();
  });

  it('renders correctly the button filter gainers', () => {
    const item = screen.getByTestId('buttonFilterGainers');
    expect(item).toBeOnTheScreen();
  });

  it('renders correctly the button filter losers', () => {
    const item = screen.getByTestId('buttonFilterLosers');
    expect(item).toBeOnTheScreen();
  });

  it('renders correctly the button filter volume24', () => {
    const item = screen.getByTestId('buttonFilterVolume24');
    expect(item).toBeOnTheScreen();
  });

  it('renders correctly the flatList component', () => {
    const item = screen.getByTestId('flatListTest');
    expect(item).toBeOnTheScreen();
  });
  it('renders correctly the modal component', () => {
    const item = screen.getByTestId('modalTest');
    expect(item).toBeOnTheScreen();
  });
});
