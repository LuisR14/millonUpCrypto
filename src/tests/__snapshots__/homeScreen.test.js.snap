// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen snapshot of screen to JSON 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        {
          "backgroundColor": "transparent",
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "marginTop": 20,
            },
            {
              "backgroundColor": "transparent",
            },
            {
              "borderRadius": 4,
              "height": undefined,
              "width": undefined,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "borderRadius": 8,
                "height": 65.71428571428571,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
                "width": 660,
              },
              [
                {
                  "overflow": "hidden",
                },
                {
                  "backgroundColor": "#E1E9EE",
                },
              ],
              {
                "borderRadius": 8,
                "height": 65.71428571428571,
                "width": 660,
              },
            ]
          }
        />
        <View
          style={
            [
              {
                "marginTop": 20,
              },
              {
                "backgroundColor": "transparent",
              },
              {
                "borderRadius": 4,
                "height": undefined,
                "width": undefined,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "marginVertical": 8,
                  "padding": 10,
                  "width": 660,
                },
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#E1E9EE",
                  },
                ],
                {
                  "borderRadius": 10,
                  "height": 131.42857142857142,
                  "width": 660,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
